documentation/src/main/asciidoc/user_guide/chapter-63-Streams.adoc

```diff
@@ -0,0 +1,416 @@
+The segments can be filtered by invoking
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#filterKeySegments-java.util.Set-[filterKeySegments]
+method on the `CacheStream`.  This is applied after the key filter above but before any intermediate operations are performed.
+
+=== Local/Invalidation/Replication Cache
+
+A stream used with a local, invalidation or replication cache can be used just the same way you would use a stream on a
+regular collection. Infinispan handles all of the translations if necessary behind the scenes and works with all
+of the more interesting options (ie. storeAsBinary, compatibility mode, and a cache loader).
+
+An example of a simple stream task is as follows
+
+==== Example
+
+Below takes a given map and assuming the values are Strings it filters all of the entries that contain the
+word JBoss and then creates a new Map with only the results of the entries that contained JBoss as a value.
+[source,java]
+----
+Map<Object, String> jbossValues = cache.entrySet().stream()
+              .filter(e -> e.getValue().contains("Jboss"))
+              .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+----
+
+=== Distribution Cache
+
+This is where streams come into their stride.  When a stream operation is performed it will send the various intermediate and
+terminal operations to each node that has pertinent data.  This allows the resulting data to be sent back to the
+originating node and intermediate values are not required to be returned which results in much better performance.
+
+
+==== Rehash Aware
+
+Internally the data is segmented and each node only performs the operations upon the data it owns as a primary owner.  This allows
+for data to be processed evenly assuming segments are granular enough to provide for equal amounts of data on each node.
+
+When you are utilizing a distributed cache data this data can be quite volatile moving between nodes when a new node is added
+or an old one has left.  When this occurs data is shuffled between nodes.  This can cause issues with processing that data as it
+can see a subset or values a second time.
+
+Distributed Streams handle this reshuffling of data automatically so you don't
+have to worry about monitoring when nodes leave or enter.  Depending on the terminal
+operation this can be done with no progress (retry), partial progress (retry
+with less required data) or keeping track of wha keys it is has processed.
+
+It is possible but highly discouraged to disable rehash awareness on the stream.  This should only be considered if
+your request can handle only seeing a subset of data if a rehash occurs.  This can be done by invoking the
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#disableRehashAware--[disableRehashAware]
+method on the `CacheStream`.  The performance gain for most operations when a rehash doesn't occur is completely
```

Comment:
You could put `CacheStream` in the link test, and make the method name monospaced as well.